OpenRA.Game/Player.cs

```diff
@@ -223,14 +223,26 @@ public override string ToString()
 		}
 
 		public Dictionary<Player, PlayerRelationship> Stances = new Dictionary<Player, PlayerRelationship>();
+
+		public PlayerRelationship RelationshipWith(Player other)
+		{
+			if (this == other)
+				return PlayerRelationship.Ally;
+
+			// Observers are considered allies to active combatants
+			if (other == null || other.Spectating)
+				return NonCombatant ? PlayerRelationship.Neutral : PlayerRelationship.Ally;
+
+			return Stances[other];
+		}
+
 		public bool IsAlliedWith(Player p)
 		{
 			// Current shroud selection is used to determine stance for spectators
-			if (p != null && p.Spectating && !NonCombatant && p.World.RenderPlayer != null)
-				return Stances[p.World.RenderPlayer] == PlayerRelationship.Ally;
+			if (p != null && p.Spectating)
```

Comment:
Shouldn't this really be `if (p == null || p.Spectating)` to match the behaviour explained in the comment? We can use just `World.RenderPlayer` without depending on `p` directly`.

Reply:
Looking at this again, I see that this then matches what `RelationshipWith` is doing -- why can't this whole method simplify to just `return RelationshipWith(p.World.RenderPlayer) == PlayerRelationship.Ally;`?

Reply:
It does desync and I don't know why at this moment. I attempt to debug that if you think it is necessary for this PR.

Reply:
i'm actually now concerned that #18077 introduced the desync and we just haven't realized it yet.